packages/itwin/reports-config-widget/src/widget/components/BulkExtractor.ts

```diff
@@ -0,0 +1,190 @@
+        states.includes(ExtractorState.Running) ? ExtractionStates.Running :
+          states.includes(ExtractorState.Succeeded) ? ExtractionStates.Succeeded : ExtractionStates.Failed;
+    return responseState;
+  }
+
+  private async getSingleState(runId: string, extractionClientApi: ExtractionClient, accessToken: string): Promise<ExtractorState> {
+    try {
+      const response = await extractionClientApi.getExtractionStatus(accessToken, runId);
+      return response.state;
+    } catch (error: any) {
+      handleError(error.status);
+    }
+    return ExtractorState.Failed;
+  }
+
+  public async startJobs(reportIds: string[]) {
+    const extractionClientApi = new ExtractionClient(
+      generateUrl(REPORTING_BASE_PATH, this._apiConfig.baseUrl)
+    );
+    const accessToken = await this._apiConfig.getAccessToken();
+
+    const extractionByIModel = new Map<string, string>();
+    for (const reportId of reportIds) {
+      const IModels = (await this.fetchReportMappings(reportId, this._apiConfig)).map((m) => m.imodelId);
+      const uniqueIModels = Array.from(new Set(IModels));
+
+      const runs: string[] = [];
+      for (const iModelId of uniqueIModels) {
+        const extraction = extractionByIModel.get(iModelId);
+        if (extraction) {
+          runs.push(extraction);
+        } else {
+          const run = await this.runExtraction(iModelId, extractionClientApi, accessToken);
+          if (run) {
+            runs.push(run.id);
+            extractionByIModel.set(iModelId, run.id);
+          }
+        }
+      }
+      this._reportRunIds.set(reportId, runs);
+    }
+  }
+
+  private async runExtraction(iModelId: string, extractionClientApi: ExtractionClient, accessToken: string) {
+    try {
+      const response = await extractionClientApi.runExtraction(
+        accessToken,
+        iModelId
+      );
+      return response;
+    } catch (error: any) {
+      handleError(error.status);
+    }
+    return undefined;
+  }
+
+  private async fetchReportMappings(
+    reportId: string,
+    apiContext: ReportsApiConfig
+  ): Promise<ReportMappingAndMapping[]> {
+    try {
+      const reportsClientApi = new ReportsClient(
+        generateUrl(REPORTING_BASE_PATH, apiContext.baseUrl)
+      );
+      const mappingsClientApi = new MappingsClient(
+        generateUrl(REPORTING_BASE_PATH, apiContext.baseUrl)
+      );
+      const accessToken = await apiContext.getAccessToken();
+      const reportMappings = await reportsClientApi.getReportMappings(
+        accessToken,
+        reportId
+      );
+      const iModelClientOptions: IModelsClientOptions = {
+        api: { baseUrl: generateUrl(Constants.api.baseUrl, apiContext.baseUrl) },
+      };
+
+      const iModelsClient: IModelsClient = new IModelsClient(iModelClientOptions);
+      const authorization =
+        AccessTokenAdapter.toAuthorizationCallback(accessToken);
+      const iModelNames = new Map<string, string>();
```

Comment:
Move the `iModelNames` cache to a private field on BulkExtractor - this will save some calls to iModels API when multiple fetches are made for report mappings. iModelNames do not change that often.

Reply:
And then write a wrapper function for that cache which would get the value from cache (Map) if it exists, and if not - get ir from iModels API and store it in the cache.

Reply:
`iModelNames` map has been moved to a private field in `BulkExtractor`

Reply:
The wrapper function for working with the iModelNames cache was not added yet.

Reply:
Function was refactored and iModelNames is no longer used.